features/mbedtls/targets/TARGET_STM/md5_alt.c

```diff
@@ -0,0 +1,124 @@
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#include "mbedtls/md5.h"
+
+#if defined(MBEDTLS_MD5_ALT)
+#include "mbedtls/platform.h"
+
+/* Implementation that should never be optimized out by the compiler */
+static void mbedtls_zeroize( void *v, size_t n ) {
+    volatile unsigned char *p = v; while( n-- ) *p++ = 0;
+}
+
+void mbedtls_md5_init( mbedtls_md5_context *ctx )
+{
+    mbedtls_zeroize( ctx, sizeof( mbedtls_md5_context ) );
+
+    /* Enable HASH clock */
+    __HAL_RCC_HASH_CLK_ENABLE();
+
+}
+
+void mbedtls_md5_free( mbedtls_md5_context *ctx )
+{
+    if( ctx == NULL )
+        return;
+
+    /* Force the HASH Periheral Clock Reset */
+    __HAL_RCC_HASH_FORCE_RESET();
+
+    /* Release the HASH Periheral Clock Reset */
+    __HAL_RCC_HASH_RELEASE_RESET();
+
+    mbedtls_zeroize( ctx, sizeof( mbedtls_md5_context ) );
+}
+
+void mbedtls_md5_clone( mbedtls_md5_context *dst,
+                        const mbedtls_md5_context *src )
+{
+    *dst = *src;
+}
+
+void mbedtls_md5_starts( mbedtls_md5_context *ctx )
+{
+    /* HASH IP initialization */
+    HAL_HASH_DeInit(&ctx->hhash_md5);
+
+    /* HASH Configuration */
+    ctx->hhash_md5.Init.DataType = HASH_DATATYPE_8B;
+    if (HAL_HASH_Init(&ctx->hhash_md5) == HAL_ERROR) {
+        // return error code
+        return;
+    }
+}
+
+void mbedtls_md5_process( mbedtls_md5_context *ctx, const unsigned char data[MBEDTLS_MD5_BLOCK_SIZE] )
+{
+    HAL_HASH_MD5_Accumulate(&ctx->hhash_md5, (uint8_t *)data, MBEDTLS_MD5_BLOCK_SIZE);
+}
+
+void mbedtls_md5_update( mbedtls_md5_context *ctx, const unsigned char *input, size_t ilen )
+{
+    size_t currentlen = ilen;
+    // store mechanism to handle MBEDTLS_MD5_BLOCK_SIZE bytes per MBEDTLS_MD5_BLOCK_SIZE bytes
+    if (currentlen == 0){ // only change HW status is size if 0
+        if(ctx->hhash_md5.Phase == HAL_HASH_PHASE_READY)
+        {
+          /* Select the MD5 mode and reset the HASH processor core, so that the HASH will be ready to compute
+             the message digest of a new message */
+          HASH->CR |= HASH_ALGOSELECTION_MD5 | HASH_CR_INIT;
+        }
+        ctx->hhash_md5.Phase = HAL_HASH_PHASE_PROCESS;
+    } else if (currentlen < (MBEDTLS_MD5_BLOCK_SIZE-ctx->sbuf_len)) {
+        // only buffurize
+        memcpy(ctx->sbuf+ctx->sbuf_len, input, currentlen);
+        ctx->sbuf_len += currentlen;
+    } else {
+        // fill buffer and process it
+        memcpy(ctx->sbuf + ctx->sbuf_len, input, (MBEDTLS_MD5_BLOCK_SIZE-ctx->sbuf_len));
+        currentlen -= (MBEDTLS_MD5_BLOCK_SIZE-ctx->sbuf_len);
+        mbedtls_md5_process(ctx, ctx->sbuf);
+        // now process every input as long as it is %4 bytes
+        size_t iter = currentlen / 4;
+        HAL_HASH_MD5_Accumulate(&ctx->hhash_md5, (uint8_t *)(input+MBEDTLS_MD5_BLOCK_SIZE-ctx->sbuf_len), (iter*4));
```

Comment:
Hello @RonEld ,
It will call 
```
  for(buffercounter = 0U; buffercounter < Size; buffercounter+=4U)
  {
    HASH->DIN = *(uint32_t*)inputaddr;
    inputaddr+=4U;
  }
```
It will do nothing in case the `Size` is `=0` (I just tested it for the 3 toolchains GCC_ARM / IAR / ARM)